consensus/core/src/core.rs

```diff
@@ -111,47 +154,196 @@ impl Core {
 
-            // emit an event that a new block is ready
-            // TODO: replace the default with the actual block ref.
-            self.signals.new_block_ready(BlockRef::default());
+            // TODO: this will be refactored once the signing path/approach has been introduced. Adding as is for now
+            // to keep things rolling in the implementation.
+            let block = Block::V1(BlockV1::new(
+                clock_round,
+                self.context.own_index,
+                now,
+                ancestors,
+                payload,
+                self.context.committee.epoch(),
+            ));
+            let signed_block = SignedBlock::new(block);
+            let verified_block = VerifiedBlock::new_verified_unserialized(signed_block)
+                .expect("Fatal error, creating a verified block failed");
+
+            //4. Add to the threshold clock
+            self.threshold_clock.add_block(verified_block.reference());
+
+            // TODO: use the DagState instead to accept the block directly. Will address on follow up PR when I'll inject
+            // the DagState. Now that's necessary to ensure that blocks aren't double processed.
+            self.block_manager.add_blocks(vec![verified_block.clone()]);
+
+            self.last_own_block = verified_block.clone();
+
+            tracing::debug!("New block created {}", verified_block);
+
+            //5. emit an event that a new block is ready
+            self.signals.new_block_ready(verified_block.reference());
+
+            return Some(verified_block);
         }
 
         None
     }
 
-    fn ready_new_block(&self) -> bool {
-        // TODO: check that we are ready to produce a new block. This will mainly check that the leader of the previous
-        // quorum exists.
-        true
+    /// Retrieves the next ancestors to propose to form a block at `clock_round` round. Also the `block_timestamp` is provided
+    /// to sanity check that everything that goes into the proposal is ensured to have a timestamp < block_timestamp
+    fn ancestors_to_propose(
+        &mut self,
+        clock_round: Round,
+        block_timestamp: BlockTimestampMs,
+    ) -> Vec<BlockRef> {
+        // Now take all the ancestors up to the clock_round (excluded) and then filter only the ones with acceptable timestamp.
+        let first_include_index = self
+            .pending_ancestors
+            .iter()
+            .position(|(block_ref, timestamp)| {
+                // We assume that our system's clock can't go backwards when we perform the check here (ex due to ntp corrections)
+                assert!(*timestamp <= block_timestamp, "Violation, ancestor block timestamp {timestamp} greater than our timestamp {block_timestamp}");
+
+                block_ref.round >= clock_round
+            })
+            .unwrap_or(self.pending_ancestors.len());
+
+        let mut taken = self.pending_ancestors.split_off(first_include_index);
```

Comment:
Unfortunately this logic cannot be taken directly. In the prototype, I can see that WAL is monotonic, and the `taken` blocks may guarantee completeness of blocks. But it does not seem to be guaranteed here.
IIUC we plan to implement this logic in or with DagState. I think the implementation should be straightforward, using the recent blocks and refs in DagState.

Reply:
Hhm actually to my understanding the monotonicity is guaranteed by two factors (1) on the way we accept blocks via the `block_manager` - all the returned results are in round order (2) the way we advance the threshold clock when blocks get accepted. Once the `threshold_clocks` advances to the next round (let's call it `clock_round`) then we know that everything that has been appended so far to the `pending_ancestors` will contain at least a quorum of `clock_round - 1` blocks and everything else before. So at least on the point of proposing for round `clock_round` we'll have the required history. I am ok to use the `DagState` instead though. Let me give that a try to see what it would look like.

Reply:
It is a bit hard to visualize the logic because some methods are stub right now. But can the following happen?
1. Accepts a block (A) at round 100, threshold clock advances to round 100.
2. Waits for leader of round 99, which gets accepted later.
3. first_include_index will be the first block at round 100, which is A. The leader block will be after first_include_index so excluded from ancestors?

Reply:
I've decided to go - at least for now (and I can explain why offline) - with a BTreeMap implementation so things can become more intuitive and avoid the edge case of having the leader block potentially missed from getting included is other (even malicious) validators manage to already produce blocks for next round. We could tackle this case by a `sort` on the vector , but keeping a BTreeMap might make things a little clearer.

Reply:
Sounds good. Btw I think it is not just malicious validator. The leader block may just propagate at different speed to different validators.

Reply:
Oh for sure, thanks for pointing out . I mostly referred to the malicious case as this would be the most severe one.